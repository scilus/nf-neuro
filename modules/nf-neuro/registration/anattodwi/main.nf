

process REGISTRATION_ANATTODWI {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(t1), path(b0), path(metric)

    output:
    tuple val(meta), path("*0GenericAffine.mat"), path("*1InverseWarp.nii.gz")  , emit: transfo_trk
    tuple val(meta), path("*1Warp.nii.gz"), path("*0GenericAffine.mat")         , emit: transfo_image
    tuple val(meta), path("*t1_warped.nii.gz")                                  , emit: t1_warped
    path "versions.yml"                                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def cpus = task.ext.cpus ? "$task.ext.cpus" : "1"

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.ext.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1
    export ANTS_RANDOM_SEED=1234

    antsRegistration --dimensionality 3 --float 0\
        --output [output,outputWarped.nii.gz,outputInverseWarped.nii.gz]\
        --interpolation Linear --use-histogram-matching 0\
        --winsorize-image-intensities [0.005,0.995]\
        --initial-moving-transform [$b0,$t1,1]\
        --transform Rigid['0.2']\
        --metric MI[$b0,$t1,1,32,Regular,0.25]\
        --convergence [500x250x125x50,1e-6,10] --shrink-factors 8x4x2x1\
        --smoothing-sigmas 3x2x1x0\
        --transform Affine['0.2']\
        --metric MI[$b0,$t1,1,32,Regular,0.25]\
        --convergence [500x250x125x50,1e-6,10] --shrink-factors 8x4x2x1\
        --smoothing-sigmas 3x2x1x0\
        --transform SyN[0.1,3,0]\
        --metric MI[$b0,$t1,1,32]\
        --metric CC[$metric,$t1,1,4]\
        --convergence [50x25x10,1e-6,10] --shrink-factors 4x2x1\
        --smoothing-sigmas 3x2x1

    mv outputWarped.nii.gz ${prefix}__t1_warped.nii.gz
    mv output0GenericAffine.mat ${prefix}__output0GenericAffine.mat
    mv output1InverseWarp.nii.gz ${prefix}__output1InverseWarp.nii.gz
    mv output1Warp.nii.gz ${prefix}__output1Warp.nii.gz


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/'
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    antsRegistration -h

    touch ${prefix}__t1_warped.nii.gz
    touch ${prefix}__output0GenericAffine.mat
    touch ${prefix}__output1InverseWarp.nii.gz
    touch ${prefix}__output1Warp.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/'
    END_VERSIONS
    """
}
