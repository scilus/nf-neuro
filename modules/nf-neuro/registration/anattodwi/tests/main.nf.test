nextflow_process {

    name "Test Process REGISTRATION_ANATTODWI"
    script "../main.nf"
    process "REGISTRATION_ANATTODWI"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/anattodwi"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    config "./nextflow.config"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "T1w.zip", "b0.zip", "dti.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - anattodwi") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                ch_t1w = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                ch_fa = ch_split_test_data.dti.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/fa.nii.gz")
                    ]
                }
                input[0] = ch_t1w
                    .join(ch_b0)
                    .join(ch_fa)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
