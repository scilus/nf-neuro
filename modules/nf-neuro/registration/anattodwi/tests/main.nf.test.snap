{
    "registration - anattodwi": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c",
                        "test__output1InverseWarp.nii.gz:md5,28694d2b41f5209ab8f1acbb219014fe"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output1Warp.nii.gz:md5,e0995852e251bee5521572a4495a0bd6",
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_warped.nii.gz:md5,9c34f424900adea263e39ed1f37a793f"
                    ]
                ],
                "3": [
                    "versions.yml:md5,4c48209a8bd74e5b5194b799152df754"
                ],
                "t1_warped": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_warped.nii.gz:md5,9c34f424900adea263e39ed1f37a793f"
                    ]
                ],
                "transfo_image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output1Warp.nii.gz:md5,e0995852e251bee5521572a4495a0bd6",
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c"
                    ]
                ],
                "transfo_trk": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c",
                        "test__output1InverseWarp.nii.gz:md5,28694d2b41f5209ab8f1acbb219014fe"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4c48209a8bd74e5b5194b799152df754"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T15:05:47.707432102"
    }
}